dependencies/src/main/java/com/google/cloud/tools/opensource/dependencies/DependencyPath.java

```diff
@@ -54,49 +71,99 @@ public boolean equals(Object o) {
     }
     
     for (int i = 0; i < path.size(); i++) {
-      if (!artifactsEqual(path.get(i), other.path.get(i))) {
+      Node thisNode = path.get(i);
+      Node otherNode = other.path.get(i);
+      if (!artifactsEqual(thisNode.getArtifact(), otherNode.getArtifact())) {
         return false; 
       }
+      if (!thisNode.getScope().equals(otherNode.getScope())) {
+        return false;
+      }
+      if (thisNode.isOptional() != otherNode.isOptional()) {
+        return false;
+      }
     }
     return true;
   }
 
-  @Override
-  public int hashCode() {
-    int hashCode = 31;
-    for (Artifact artifact : path) {
-      hashCode = 37 * hashCode
-          + (artifact.getGroupId() + ":" + artifact.getArtifactId() + ":" + artifact.getVersion())
-              .hashCode();
-    }
-    return hashCode; 
-  }
-  
   /**
-   * Artifacts are considered to be the same if they have the same group ID,
-   * artifact ID, and version.
+   * Artifacts are considered to be the same if they have the same group ID, artifact ID, and
+   * version.
+   *
+   * <p>{@link org.eclipse.aether.artifact.DefaultArtifact#equals(Object)} uses {@code file} field,
+   * which is unnecessary for our usage.
    */
   private static boolean artifactsEqual(Artifact artifact1, Artifact artifact2) {
     return artifact1.getArtifactId().equals(artifact2.getArtifactId())
         && artifact1.getGroupId().equals(artifact2.getGroupId())
         && artifact1.getVersion().equals(artifact2.getVersion());
   }
 
-  public int size() {
-    return path.size();
+  @Override
+  public int hashCode() {
+    int hashCode = 31;
+    for (Node node : path) {
+      Artifact artifact = node.getArtifact();
+      hashCode =
+          37 * hashCode
+              + Objects.hash(
+                  artifact.getGroupId(),
+                  artifact.getArtifactId(),
+                  artifact.getVersion(),
+                  node.scope,
+                  node.optional);
+    }
+    return hashCode;
   }
 
-  public Artifact getLeaf() {
-    return path.get(size() - 1);
-  }
+  /** Node in a dependency tree, holding scope and optional flag. */
+  private static class Node {
+    private Artifact artifact;
+    private String scope;
+    private boolean optional;
 
-  public List<Artifact> getPath() {
-    return path;
-  }
+    private Node(Artifact artifact, String scope, boolean optional) {
+      this.artifact = checkNotNull(artifact);
+      this.scope = checkNotNull(scope);
+      this.optional = optional;
+    }
 
-  // TODO think about index out of bounds
```

Comment:
This TODO is removed; I don't see this array index access is special. If out of bounds, let JVM raise IndexOutOfBoundsException.